File: client/Core/VlessUri.cs
using System.Web;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace VlessMonitor.Core;

/// <summary>
/// Parses a vless:// URI and generates an xray-core client config.
/// </summary>
public class VlessUri
{
    public string Uuid = "";
    public string Host = "";
    public int Port = 443;
    public string Name = "";
    public string Network = "tcp";
    public string Encryption = "none";
    public string Security = "none";
    public string Flow = "";
    public string PublicKey = "";
    public string Fingerprint = "chrome";
    public string Sni = "";
    public string ShortId = "";
    public string SpiderX = "/";

    public static VlessUri Parse(string uri)
    {
        if (!uri.StartsWith("vless://"))
            throw new ArgumentException("Не VLESS-ссылка (должна начинаться с vless://)");

        var u = new Uri(uri);
        var q = HttpUtility.ParseQueryString(u.Query);

        return new VlessUri
        {
            Uuid = u.UserInfo,
            Host = u.Host,
            Port = u.Port > 0 ? u.Port : 443,
            Name = Uri.UnescapeDataString(u.Fragment.TrimStart('#')),
            Network = q["type"] ?? "tcp",
            Encryption = q["encryption"] ?? "none",
            Security = q["security"] ?? "none",
            Flow = q["flow"] ?? "",
            PublicKey = q["pbk"] ?? "",
            Fingerprint = q["fp"] ?? "chrome",
            Sni = q["sni"] ?? "",
            ShortId = q["sid"] ?? "",
            SpiderX = Uri.UnescapeDataString(q["spx"] ?? "/"),
        };
    }

    /// <summary>Build a minimal xray-core config JSON for this VLESS client.</summary>
    public string BuildXrayConfig(int socksPort)
    {
        var streamSettings = new JsonObject { ["network"] = Network };

        if (Security == "reality")
        {
            streamSettings["security"] = "reality";
            streamSettings["realitySettings"] = new JsonObject
            {
                ["serverName"] = Sni,
                ["fingerprint"] = Fingerprint,
                ["shortId"] = ShortId,
                ["publicKey"] = PublicKey,
                ["spiderX"] = SpiderX,
                ["show"] = false,
            };
        }
        else if (Security == "tls")
        {
            streamSettings["security"] = "tls";
            streamSettings["tlsSettings"] = new JsonObject
            {
                ["serverName"] = Sni,
                ["fingerprint"] = Fingerprint,
            };
        }

        var user = new JsonObject { ["id"] = Uuid, ["encryption"] = Encryption };
        if (!string.IsNullOrEmpty(Flow)) user["flow"] = Flow;

        var config = new JsonObject
        {
            ["log"] = new JsonObject { ["loglevel"] = "warning" },
            ["inbounds"] = new JsonArray
            {
                new JsonObject
                {
                    ["listen"] = "127.0.0.1",
                    ["port"] = socksPort,
                    ["protocol"] = "socks",
                    ["settings"] = new JsonObject { ["auth"] = "noauth", ["udp"] = true },
                    ["tag"] = "socks-in",
                }
            },
            ["outbounds"] = new JsonArray
            {
                new JsonObject
                {
                    ["tag"] = "vless-out",
                    ["protocol"] = "vless",
                    ["settings"] = new JsonObject
                    {
                        ["vnext"] = new JsonArray
                        {
                            new JsonObject
                            {
                                ["address"] = Host,
                                ["port"] = Port,
                                ["users"] = new JsonArray { user },
                            }
                        }
                    },
                    ["streamSettings"] = streamSettings,
                },
                new JsonObject { ["tag"] = "direct", ["protocol"] = "freedom" },
            },
        };

        return config.ToJsonString(new JsonSerializerOptions { WriteIndented = true });
    }
}

File: vless_uri.py
"""
Parse a VLESS URI into components and generate an xray-core outbound config.

vless://UUID@host:port?type=tcp&security=reality&pbk=KEY&fp=chrome&sni=SNI&sid=ID&spx=PATH#name
"""
import json
import urllib.parse
from dataclasses import dataclass, field


@dataclass
class VlessConfig:
    uuid: str
    host: str
    port: int
    name: str = ""
    network: str = "tcp"
    encryption: str = "none"
    security: str = "none"
    flow: str = ""
    # Reality settings
    public_key: str = ""
    fingerprint: str = "chrome"
    sni: str = ""
    short_id: str = ""
    spider_x: str = "/"


def parse_vless_uri(uri: str) -> VlessConfig:
    if not uri.startswith("vless://"):
        raise ValueError("Not a vless:// URI")

    parsed = urllib.parse.urlparse(uri)
    params = dict(urllib.parse.parse_qsl(parsed.query))

    return VlessConfig(
        uuid=parsed.username or "",
        host=parsed.hostname or "",
        port=parsed.port or 443,
        name=urllib.parse.unquote(parsed.fragment or ""),
        network=params.get("type", "tcp"),
        encryption=params.get("encryption", "none"),
        security=params.get("security", "none"),
        flow=params.get("flow", ""),
        public_key=params.get("pbk", ""),
        fingerprint=params.get("fp", "chrome"),
        sni=params.get("sni", ""),
        short_id=params.get("sid", ""),
        spider_x=urllib.parse.unquote(params.get("spx", "/")),
    )


def build_xray_client_config(vless: VlessConfig, socks_port: int = 10808) -> dict:
    """Generate a minimal xray-core config for a VLESS client."""

    stream_settings: dict = {"network": vless.network}

    if vless.security == "reality":
        stream_settings["security"] = "reality"
        stream_settings["realitySettings"] = {
            "serverName": vless.sni,
            "fingerprint": vless.fingerprint,
            "shortId": vless.short_id,
            "publicKey": vless.public_key,
            "spiderX": vless.spider_x,
            "show": False,
        }
    elif vless.security == "tls":
        stream_settings["security"] = "tls"
        stream_settings["tlsSettings"] = {
            "serverName": vless.sni,
            "fingerprint": vless.fingerprint,
        }

    user = {"id": vless.uuid, "encryption": vless.encryption}
    if vless.flow:
        user["flow"] = vless.flow

    return {
        "log": {"loglevel": "warning"},
        "inbounds": [
            {
                "listen": "127.0.0.1",
                "port": socks_port,
                "protocol": "socks",
                "settings": {"auth": "noauth", "udp": True},
                "tag": "socks-in",
            }
        ],
        "outbounds": [
            {
                "tag": "vless-out",
                "protocol": "vless",
                "settings": {
                    "vnext": [
                        {
                            "address": vless.host,
                            "port": vless.port,
                            "users": [user],
                        }
                    ]
                },
                "streamSettings": stream_settings,
            },
            {"tag": "direct", "protocol": "freedom"},
        ],
    }
